### ISSUE TITLE
Namespace design

### ISSUE THREAD

**Issue by nguerrera** [Azure/adl#234]
Today we have interfaces which are bags of operations, but there is no hierarchy to them. 

This could be augmented or replaced with namespaces which could be hierarchical as they are in other languages.

The hierarchy of namespaces could align with the path hierarchy for REST API definition.

For example, GET /foo/bar/baz could something like bet

```
namespace foo {  
  namespace bar {
     @get baz():string
  }
}
```

### Some design points to consider:

* Namespaces should probably be "partial" like in other languages, meaning you can spread its member namespaces, operations across more than one source file.

* Should models go inside namespaces too?

* Do we need to retain interfaces at all or is it just that namespaces can have functions?

* How do namespaces interact with import/export/package management (TODO: issue link)

**Comment by nguerrera** [comment]
Notes from chat with @bterlson:
* implicit foo/bar/baz is "cute", go with @uri (name maybe TBD) + path on each namespace

**Comment by daviwil** [comment]
So far I like this a lot!  Here are some general thoughts:

- How will namespaces get "sourced" from extension packages?  If we're using a plain `using Some.Namespace`, do we assume that all of the extension packages are pulled in and added to a "load path" of sorts?

- Can we be more explicit about pulling in JS files to import decorator code into a specific namespace?  Right now we're just pulling in whatever .js files are in the folder path, but maybe it's better to be more explicit?  `import` could possibly be used for this:  `import './decorators.js'`.  If you use this statement inside of a sub-namespace, the imported decorators get added there.

One interesting aspect of this (which I've seen discussed in other issues here) is that we can use the root namespace to determine the service title and metadata when emitting Swagger:

```
@title("My Service Client")
namespace MyServiceClient {
}
```

Now the ADL user can refer to a specific namespace when compiling the output:

```
adl compile --namespace MyServiceClient samples/myserviceclient
```

**Comment by nguerrera** [comment]
> How will namespaces get "sourced" from extension packages? If we're using a plain using Some.Namespace, do we assume that all of the extension packages are pulled in and added to a "load path" of sorts?

I don't know, except that at a high level somewhere I think you specify which packages (and versions!) you are referencing and somehow the compiler gets enough info to get to their source make their definitions available as described here. I think ironing out the mechanics is up to #236 to define, but not saying package names in the source is something that @bterslon and I discussed and we think it will be more productive. C# also works this way. That said, I suspect we will see a strong correspondence between packages and namespaces as you see in .NET.

> import './decorators.js'. If you use this statement inside of a sub-namespace, the imported decorators get added there.

This is interesting. Another idea I thought about was to have some sort of decorator declaration syntax (like a forward declaration in C or an extern method in C#)  that also tells you the typing for the decorator, and gives you a place do document them in the language (OMG, decorated decorators?).
```
namespace MyLib {
    decorator MyDecorator(description: string); 
 }
 ```

These aren't mutually exclusive as there are three questions: 1) What's the namespace? 2) What's the signature 3) Where do I get the JS, and this just answers one and 2. We could do 3 through a separate import or combine it with above. 

Or maybe there's a way to keep it all magic like a folder convention or something by namespace off of each imported package and in current project?

**Comment by bterlson** [comment]
Overall I like this proposal a lot! Couple thoughts:

* I think we'll need a keyword to kick off the rpc endpoint, otherwise I think it needs to be treated like a special kind of expression in the parser (we'd probably try parsing an expression, and if we get an identifier and the next token is openparen, actually it's a statement). We leave ourselves open to future ambiguity. I'd suggest something like `fn` or `rpc`.
* I think we can leave the `using` alias form for now. It should be easy to add later when we need it, and hopefully we'll have more of the language done at that point. I'm worried about the conflict with `model =`, and `model =` is on thin ice as far as I'm concerned.
* Are we sure we will have a complete picture of what's in a namespace before running the decorator on it? I can imagine e.g. wanting to iterate over the contents of a namespace.
* We could potentially use `namespace global { }` as a special form to append decorators onto the global namespace without taking up more syntactic space...

**Comment by nguerrera** [comment]
> I think we'll need a keyword to kick off the rpc endpoint

👍 I knew there was something from our chats long ago about namespaces that I forgot and I think this was it!

I *think *the current grammar is still easy to parse because we reserved the tokens of the other statements as keywords, but but regardless that probably won't hold up over time because we'll inevitably want to add contextual keywords for compat at some point. So that's moot. Do we have a preference for the keyword. I'm usually averse to abbreviations but `fn` has plenty of prior art and you're going to have a lot of them so being terse does feel worth it here. I find `rpc` a little weird for some reason. Just to throw out another idea, how do we feel about calling these "operations"? op keyword? That maps to vocabulary in swagger and autorest, but I think I still like "function" better for the core language concept.

**Comment by nguerrera** [comment]
> Are we sure we will have a complete picture of what's in a namespace before running the decorator on it? I can imagine e.g. wanting to iterate over the contents of a namespace.

This is a great question and I am not sure. I think I need to go back and study decorator invocation. I think having a speclet for this would be useful so we can update it like the grammar with proposals like this. I can try to write that, but will likely need your help.

**Comment by nguerrera** [comment]
> I think we can leave the using alias form for now. It should be easy to add later when we need it, and hopefully we'll have more of the language done at that point. I'm worried about the conflict with model =, and model = is on thin ice as far as I'm concerned.

Agreed. I'll cut it.

**Comment by nguerrera** [comment]
> We could potentially use namespace global { } as a special form to append decorators onto the global namespace without taking up more syntactic space...

I like this. Or maybe `namespace {`. On second thought anonymous namespace and global namespace are two very different things in C++ and this is the anonymous namespace syntax. There is a `global` concept in C# that's related to this. OK, I like `namespace global` 😀.

**Comment by bterlson** [comment]
Love the `op` keyword!! Makes me think... should we use `ns` instead of `namespace` too?

**Comment by nguerrera** [comment]
OK, I'll go with `op`. 😀

I'm not crazy about `ns`. I feel like this is not frequent enough to need abbreviation, and `namespace` is in C#, C++, and TypeScript with the proposed syntax.

**Comment by nguerrera** [comment]
Description and #274 updated.

**Comment by nguerrera** [comment]
I put the `namespace global` idea in option 5 of #232

**Comment by jamesc** [comment]
1. How does this overlap with the `Service` spec.  Can we embed namespaces inside Service, or only an `op`.
2. How do decorators on namespaces work?  If I have two `op`s in different files but the same namespace, and I set decorators on both namespaces.

**Comment by jamesc** [comment]
> OK, I'll go with `op`. 😀

Smithy has `operation`, gRPC has `rpc`.  I like `op`.

**Comment by nguerrera** [comment]
> How does this overlap with the Service spec. Can we embed namespaces inside Service, or only an op.

@jamesc Great question. If we add `service`, I think we could still decide either way. One of the options in the service spec is to forego `service` altogether and add metadata to `namespace` to describe service metadata. There's still the separate question of more than one service per "project". If we want exactly one, then we could say `namespace global` is the service definition, otherwise we could say that you can define multiple services with multiple namespaces. Similarly, if we do add `service` we can make the same decision independently: just one `service` outside namespaces, or you can put different `services` in different namespaces in the same "project". Based on this thinking, I think we can safely approve this proposal without answering that question now and leave it to #232 to answer. Would you agree?

**Comment by nguerrera** [comment]
> How do decorators on namespaces work? If I have two ops in different files but the same namespace, and I set decorators on both namespaces.

@jamesc, I attempted to answer this with the following, but I wasn't very precise:

> * Namespaces are "partial"
>    * Decorators on namespace and namespace members are merged together across source files

I meant "partial" here in the sense of C# `partial class`, meaning that if you do what you're asking you get 1 namespace entity with the combined set of decorators from both source files, and the 2 ops as its children.

At a high level, I think this is what we want, but @bterlson alluded to a challenge in this space:

> Are we sure we will have a complete picture of what's in a namespace before running the decorator on it? 

I still don't have an answer to that yet, and I think I'd need enough of an implementation to have a playground for experimentation around this to come up with good answers here.

**Comment by bterlson** [comment]
I think James's question brings in another open design space we have which is, what is our re-opening strategy generally? It is a little strange that I can just add a decorator to an entire namespace and start doing things with declarations from other parts of my API or even in a library. Consider:

```
// inside the RPaaS library
namespace Azure.RPaaS {
  model Resource<T> { ... }
}

// inside some user ADL file
@doFunThings
namespace Azure.RPaaS { };
```

But I suppose this is the moral equivalent of appending stuff into this namespace? I guess I'm not familiar with namespacing languages to have a good feel for how gross this is, really.

**Comment by daviwil** [comment]
It seems like reopening would be useful for attaching protocol-specific metadata or customizations that someone needs to make to an existing ADL spec.  Basically, imagine using reopening for this instead of AutoRest's transform feature (which is pretty hard to use).

However, this kind of reopening could definitely introduce behavior that is hard to trace or even prevent with linting, so it seems like a Pandora's Box kind of situation.

**Comment by jamesc** [comment]
> Similarly, if we do add `service` we can make the same decision independently: just one `service` outside namespaces, or you can put different `services` in different namespaces in the same "project". Based on this thinking, I think we can safely approve this proposal without answering that question now and leave it to #232 to answer. Would you agree?

Yep, I don't think it's a blocker and we'll have to rework the grammer either way if we add service.

**Comment by nguerrera** [comment]
The other namespacing languages I'm familiar with leave namespaces open and you can keep appending members to them. But, they don't have annotations on namespaces. For example, you can't put an attribute on a C# namespace. The other open things besides namespace in C# have the keyword `partial` explicitly to say somebody else may be contributing to this. You can add attributes to a `partial` class or a `partial` method in C#:

![image](https://user-images.githubusercontent.com/75470/106961397-60f94600-66f2-11eb-87c6-f7c78a5df3d7.png)

**Comment by nguerrera** [comment]
Of course, C# attributes are just data added to the output of compilation. Whereas our decorators run during compilation and have side effects so it's a lot more complex to spread them out: what order do they run in, etc.

**Comment by nguerrera** [comment]
We could go in the opposite direction such that there is different syntax for a single canonical namespace definition and for adding more members to a namespace defined elsewhere. But then that strictness seems the opposite of our desire to allow sidecar decoration somehow. If we have sidecar decoration in any form, I don't think it's any less problematic than what you can do with namespaces here, which is not to say it isn't problematic. 😁

**Comment by nguerrera** [comment]
Getting off topic and we should probably make a sidecar decoration issue, but I'm wondering if we had two different things: data annotations and macro-like decorators (what we have now). The data could be spread out, but the macros have to go with a single canonical definition. The macros can consume the data which is parsed and merged before the macros run.

**Comment by bterlson** [comment]
If we want re-opening, global side effects, and reflection, it feels like we need to put users in control of the order of evaluation, so e.g. you can reorder side effects when needed, or say "define this service, apply these behaviors, then apply these docs" in an explicit order.

**Comment by nguerrera** [comment]
We could do `import/include/load  <file>` in addition to `using <namespace>`. That's basically how C++ works (or worked, I'm very behind on modern C++. 😁) Instead of a directory, you'd give the root file to start with. Import/include/load puts everything behind full namespace qualification, using brings a namespaces members into the current lexical scope without qualification.

Like 

C++
``` C++
#include "tech.h"
using namespace Tech; // or just use Tech::Name everywhere
```

Possible ADL
```
import "tech.adl"
using Tech; // or just use Tech.Name everywhere
```

Just for another option, F# requires an ordering of files and the mechanism is to re-order them in your "project file" so that they are passed to the compiler in an order.

I really want to keep a way to say "bring in everything in this namespace into scope so that I don't have to qualify it". I think we can have that and explicit ordering together. 

(Side rant: Even with explicit ordering, something still feels off to me. I hope that we develop patterns that don't make it common to have to think about it too much. I think we're trying to give a declarative feel with our language and decorators, and the more it turns into debugging an imperative program on the regular, the less this "description" language will be easy to use. It's one thing to have the declare before use ordering of C++ and F#, but quite another to be mixing together a bunch of files with side effects and needing to reorder them based on bad interactions. This reminds me of msbuild where your project looks like descriptive xml, but is really a program with some very peculiar semantics.  I *really* want to avoid the experience becoming as frustrating as that can be. Anyway, nothing actionable here, just venting a little about some of the risks ahead... In fairness, there are much better examples of description languages that are actually imperative than msbuild. Lisp, Ruby, and LUA-internal DSLs come to mind. Those have the advantage that you're inside a full featured language so you can always break out of jail when the abstraction leaks. https://premake.github.io/ is a good example that I enjoyed in a past life.)

Anyway, is there a way we can split up the work here? For example, could we say on first pass at namespaces, it is simply an error to decorate a namespace in more than one file. Or even make namespaces closed like models and interfaces are now at first: you have to put everything in the namespace in the same file in the same declaration. Then have a separate work item to track opening namespaces and/or side-car decoration in general? Just wondering about how to get the most breaking parts in ASAP without necessarily solving this hard problem first.

**Comment by nguerrera** [comment]
Just to update here, #280 makes replaces `interface` with `namespace` and adds the `op` keyword for operation members of namespace. The rest of the changes once that is absorbed should be non-breaking.
